# PyCitySchools 

In [1]:
# Dependencies and Setup
import pandas as pd
import numpy as np


In [2]:
# File to Load 
school_data_to_load = "Resources/schools_complete.csv"
student_data_to_load = "Resources/students_complete.csv"


In [3]:
# Read School and Student Data File and store into Pandas Data Frames
school_data = pd.read_csv(school_data_to_load, encoding="utf-8")
school_data.head()

School ID            school_name      type  size   budget
0          0      Huang High School  District  2917  1910635
1          1   Figueroa High School  District  2949  1884411
2          2    Shelton High School   Charter  1761  1056600
3          3  Hernandez High School  District  4635  3022020
4          4    Griffin High School   Charter  1468   917500

In [4]:
student_data = pd.read_csv(student_data_to_load, encoding="utf-8")
student_data.head()

Student ID       student_name gender grade        school_name  \
0           0       Paul Bradley      M   9th  Huang High School   
1           1       Victor Smith      M  12th  Huang High School   
2           2    Kevin Rodriguez      M  12th  Huang High School   
3           3  Dr. Richard Scott      M  12th  Huang High School   
4           4         Bonnie Ray      F   9th  Huang High School   

   reading_score  math_score  
0             66          79  
1             94          61  
2             90          60  
3             67          58  
4             97          84

In [5]:
# Combine the data into a single dataset
school_data_complete = pd.merge(student_data, school_data, how="left", on=["school_name", "school_name"])
school_data_complete.head()

Student ID       student_name gender grade        school_name  \
0           0       Paul Bradley      M   9th  Huang High School   
1           1       Victor Smith      M  12th  Huang High School   
2           2    Kevin Rodriguez      M  12th  Huang High School   
3           3  Dr. Richard Scott      M  12th  Huang High School   
4           4         Bonnie Ray      F   9th  Huang High School   

   reading_score  math_score  School ID      type  size   budget  
0             66          79          0  District  2917  1910635  
1             94          61          0  District  2917  1910635  
2             90          60          0  District  2917  1910635  
3             67          58          0  District  2917  1910635  
4             97          84          0  District  2917  1910635

## District Summary

* Calculate the total number of schools

* Calculate the total number of students

* Calculate the total budget

* Calculate the average math score 

* Calculate the average reading score

* Calculate the overall passing rate (overall average score), i.e. (avg. math score + avg. reading score)/2

* Calculate the percentage of students with a passing math score (70 or greater)

* Calculate the percentage of students with a passing reading score (70 or greater)

* Create a dataframe to hold the above results

* Optional: give the displayed data cleaner formatting

In [6]:
# View column names
school_data_complete.columns

Index(['Student ID', 'student_name', 'gender', 'grade', 'school_name',
       'reading_score', 'math_score', 'School ID', 'type', 'size', 'budget'],
      dtype='object')

In [7]:
#Calculate the total number of schools

total_school = len(school_data_complete['school_name'].unique())

total_school

15

In [8]:
#Calculate the total number of students
total_student_no = len(school_data_complete['Student ID'].unique())
total_student_no

39170

In [9]:
#Calculate the total budget
total_budget = school_data['budget'].sum()
total_budget

24649428

In [10]:
#Calculate the average math score
avg_math_score = school_data_complete['math_score'].mean()
avg_math_score

78.98537145774827

In [11]:
#Calculate the average reading score
avg_read_score = school_data_complete['reading_score'].mean()
avg_read_score

81.87784018381414

In [12]:
#Calculate the overall passing rate (overall average score), i.e. (avg. math score + avg. reading score)/2
passing_rate = (avg_math_score + avg_read_score)/2
passing_rate

80.43160582078121

In [13]:
# Calculate the percentage of students with a passing math score (70 or greater)
pass_math = (school_data_complete.loc[school_data_complete['math_score']>=70])
pass_math_percent = (len(pass_math['math_score'])/total_student_no) * 100
pass_math_percent


74.9808526933878

In [14]:
pass_math.head()

Student ID   student_name gender grade        school_name  reading_score  \
0           0   Paul Bradley      M   9th  Huang High School             66   
4           4     Bonnie Ray      F   9th  Huang High School             97   
5           5  Bryan Miranda      M   9th  Huang High School             94   
6           6  Sheena Carter      F  11th  Huang High School             82   
8           8   Michael Roth      M  10th  Huang High School             95   

   math_score  School ID      type  size   budget  
0          79          0  District  2917  1910635  
4          84          0  District  2917  1910635  
5          94          0  District  2917  1910635  
6          80          0  District  2917  1910635  
8          87          0  District  2917  1910635

In [15]:
# Calculate the percentage of students with a passing reading score (70 or greater)
pass_read = (school_data_complete.loc[school_data_complete['reading_score']>=70])
pass_read_percent = ((len(pass_read['reading_score'])/total_student_no) * 100)
pass_read_percent

85.80546336482001

In [16]:
pass_read.head()

Student ID     student_name gender grade        school_name  reading_score  \
1           1     Victor Smith      M  12th  Huang High School             94   
2           2  Kevin Rodriguez      M  12th  Huang High School             90   
4           4       Bonnie Ray      F   9th  Huang High School             97   
5           5    Bryan Miranda      M   9th  Huang High School             94   
6           6    Sheena Carter      F  11th  Huang High School             82   

   math_score  School ID      type  size   budget  
1          61          0  District  2917  1910635  
2          60          0  District  2917  1910635  
4          84          0  District  2917  1910635  
5          94          0  District  2917  1910635  
6          80          0  District  2917  1910635

In [17]:
# Create a dataframe to hold the above results
district_df = pd.DataFrame({
    "Total Schools": [total_school],
    "Total Students": [total_student_no], 
    "Total Budget": [total_budget],
    "Average Math Score": [avg_math_score],
    "Average Reading Score": [avg_read_score],
    "Overall Passing Rate": [passing_rate],
    "Passing Math Score %": [pass_math_percent],
    "Passing Reading Score %": [pass_read_percent]
})

# Optional: give the displayed data cleaner formatting

district_df["Total Budget"] = district_df["Total Budget"].map("${:,.2f}".format)
district_df["Average Math Score"] = district_df["Average Math Score"].map("{:.2f}".format)
district_df["Average Reading Score"] = district_df["Average Reading Score"].map("{:.2f}".format)
district_df["Overall Passing Rate"] = district_df["Overall Passing Rate"].map("{:.2f}".format)
district_df["Passing Math Score %"] = district_df["Passing Math Score %"].map("{:.2f}".format)
district_df["Passing Reading Score %"] = district_df["Passing Reading Score %"].map("{:.2f}".format)
district_df

Total Schools  Total Students    Total Budget Average Math Score  \
0             15           39170  $24,649,428.00              78.99   

  Average Reading Score Overall Passing Rate Passing Math Score %  \
0                 81.88                80.43                74.98   

  Passing Reading Score %  
0                   85.81

## School Summary

* Create an overview table that summarizes key metrics about each school, including:
  * School Name
  * School Type
  * Total Students
  * Total School Budget
  * Per Student Budget
  * Average Math Score
  * Average Reading Score
  * % Passing Math
  * % Passing Reading
  * Overall Passing Rate (Average of the above two)
  
* Create a dataframe to hold the above results

In [40]:
school_summary = school_data_complete.groupby(['school_name'])
sch_type = school_summary['type'].first()
sch_students = school_summary['Student ID'].count()
sch_budget = school_summary['budget'].first()
per_student_budget = sch_budget/sch_students
sch_avg_math = school_summary['math_score'].mean()
sch_avg_read = school_summary['reading_score'].mean()

In [41]:
sch_pass_math = pass_math.groupby(['school_name'])
sch_pass_math_percent = (sch_pass_math['math_score'].count()/sch_students) * 100
sch_pass_read = pass_read.groupby(['school_name'])
sch_pass_read_percent = (sch_pass_read['reading_score'].count()/sch_students) * 100
sch_overall = (sch_pass_math_percent + sch_pass_read_percent)/2


In [42]:
#unformatted
sch_summary_0 = pd.DataFrame({
    "School Type": sch_type, 
    "Total Students": sch_students,
    "Total School Budget": sch_budget,
    "Per Student Budget": per_student_budget,
    "Average Math Score": sch_avg_math,
    "Average Reading Score": sch_avg_read,
    "% Passing Math": sch_pass_math_percent,
    "% Passing Reading":sch_pass_read_percent,
    "Overall Passing Rate": sch_overall
})


In [43]:
sch_summary_df = pd.DataFrame({
    "School Type": sch_type, 
    "Total Students": sch_students,
    "Total School Budget": sch_budget,
    "Per Student Budget": per_student_budget,
    "Average Math Score": sch_avg_math,
    "Average Reading Score": sch_avg_read,
    "% Passing Math": sch_pass_math_percent,
    "% Passing Reading":sch_pass_read_percent,
    "Overall Passing Rate": sch_overall
})

sch_summary_df["Total Students"] = sch_summary_df["Total Students"].map("{:,}".format)
sch_summary_df["Total School Budget"] = sch_summary_df["Total School Budget"].map("${:,.2f}".format)
sch_summary_df["Per Student Budget"] = sch_summary_df["Per Student Budget"].map("${:,.2f}".format)
sch_summary_df["Average Math Score"] = sch_summary_df["Average Math Score"].map("{:.2f}".format)
sch_summary_df["Average Reading Score"] = sch_summary_df["Average Reading Score"].map("{:.2f}".format)
sch_summary_df["% Passing Math"] = sch_summary_df["% Passing Math"].map("{:.2f}%".format)
sch_summary_df["% Passing Reading"] = sch_summary_df["% Passing Reading"].map("{:.2f}%".format)
sch_summary_df["Overall Passing Rate"] = sch_summary_df["Overall Passing Rate"].map("{:.2f}%".format)
sch_summary_df

School Type Total Students Total School Budget  \
school_name                                                            
Bailey High School       District          4,976       $3,124,928.00   
Cabrera High School       Charter          1,858       $1,081,356.00   
Figueroa High School     District          2,949       $1,884,411.00   
Ford High School         District          2,739       $1,763,916.00   
Griffin High School       Charter          1,468         $917,500.00   
Hernandez High School    District          4,635       $3,022,020.00   
Holden High School        Charter            427         $248,087.00   
Huang High School        District          2,917       $1,910,635.00   
Johnson High School      District          4,761       $3,094,650.00   
Pena High School          Charter            962         $585,858.00   
Rodriguez High School    District          3,999       $2,547,363.00   
Shelton High School       Charter          1,761       $1,056,600.00   
Thomas High School        Charter          1,635       $1,043,130.00   
Wilson High School        Charter          2,283       $1,319,574.00   
Wright High School        Charter          1,800       $1,049,400.00   

                      Per Student Budget Average Math Score  \
school_name                                                   
Bailey High School               $628.00              77.05   
Cabrera High School              $582.00              83.06   
Figueroa High School             $639.00              76.71   
Ford High School                 $644.00              77.10   
Griffin High School              $625.00              83.35   
Hernandez High School            $652.00              77.29   
Holden High School               $581.00              83.80   
Huang High School                $655.00              76.63   
Johnson High School              $650.00              77.07   
Pena High School                 $609.00              83.84   
Rodriguez High School            $637.00              76.84   
Shelton High School              $600.00              83.36   
Thomas High School               $638.00              83.42   
Wilson High School               $578.00              83.27   
Wright High School               $583.00              83.68   

                      Average Reading Score % Passing Math % Passing Reading  \
school_name                                                                    
Bailey High School                    81.03         66.68%            81.93%   
Cabrera High School                   83.98         94.13%            97.04%   
Figueroa High School                  81.16         65.99%            80.74%   
Ford High School                      80.75         68.31%            79.30%   
Griffin High School                   83.82         93.39%            97.14%   
Hernandez High School                 80.93         66.75%            80.86%   
Holden High School                    83.81         92.51%            96.25%   
Huang High School                     81.18         65.68%            81.32%   
Johnson High School                   80.97         66.06%            81.22%   
Pena High School                      84.04         94.59%            95.95%   
Rodriguez High School                 80.74         66.37%            80.22%   
Shelton High School                   83.73         93.87%            95.85%   
Thomas High School                    83.85         93.27%            97.31%   
Wilson High School                    83.99         93.87%            96.54%   
Wright High School                    83.95         93.33%            96.61%   

                      Overall Passing Rate  
school_name                                 
Bailey High School                  74.31%  
Cabrera High School                 95.59%  
Figueroa High School                73.36%  
Ford High School                    73.80%  
Griffin High School                 95.27%  
Hernandez High School               73.81%  
Hold

In [44]:
sch_summary_df.dtypes

School Type              object
Total Students           object
Total School Budget      object
Per Student Budget       object
Average Math Score       object
Average Reading Score    object
% Passing Math           object
% Passing Reading        object
Overall Passing Rate     object
dtype: object

## Top Performing Schools (By Passing Rate)

* Sort and display the top five schools in overall passing rate

In [45]:
sort_overall = sch_summary_df.sort_values(["Overall Passing Rate"], ascending=False)
sort_overall.reset_index
sort_overall.head(5)


School Type Total Students Total School Budget  \
school_name                                                          
Cabrera High School     Charter          1,858       $1,081,356.00   
Thomas High School      Charter          1,635       $1,043,130.00   
Griffin High School     Charter          1,468         $917,500.00   
Pena High School        Charter            962         $585,858.00   
Wilson High School      Charter          2,283       $1,319,574.00   

                    Per Student Budget Average Math Score  \
school_name                                                 
Cabrera High School            $582.00              83.06   
Thomas High School             $638.00              83.42   
Griffin High School            $625.00              83.35   
Pena High School               $609.00              83.84   
Wilson High School             $578.00              83.27   

                    Average Reading Score % Passing Math % Passing Reading  \
school_name                                                                  
Cabrera High School                 83.98         94.13%            97.04%   
Thomas High School                  83.85         93.27%            97.31%   
Griffin High School                 83.82         93.39%            97.14%   
Pena High School                    84.04         94.59%            95.95%   
Wilson High School                  83.99         93.87%            96.54%   

                    Overall Passing Rate  
school_name                               
Cabrera High School               95.59%  
Thomas High School                95.29%  
Griffin High School               95.27%  
Pena High School                  95.27%  
Wilson High School                95.20%

## Bottom Performing Schools (By Passing Rate)

* Sort and display the five worst-performing schools

In [46]:
sort_overall = sch_summary_df.sort_values(["Overall Passing Rate"])
sort_overall.reset_index
sort_overall.head(5)

School Type Total Students Total School Budget  \
school_name                                                            
Rodriguez High School    District          3,999       $2,547,363.00   
Figueroa High School     District          2,949       $1,884,411.00   
Huang High School        District          2,917       $1,910,635.00   
Johnson High School      District          4,761       $3,094,650.00   
Ford High School         District          2,739       $1,763,916.00   

                      Per Student Budget Average Math Score  \
school_name                                                   
Rodriguez High School            $637.00              76.84   
Figueroa High School             $639.00              76.71   
Huang High School                $655.00              76.63   
Johnson High School              $650.00              77.07   
Ford High School                 $644.00              77.10   

                      Average Reading Score % Passing Math % Passing Reading  \
school_name                                                                    
Rodriguez High School                 80.74         66.37%            80.22%   
Figueroa High School                  81.16         65.99%            80.74%   
Huang High School                     81.18         65.68%            81.32%   
Johnson High School                   80.97         66.06%            81.22%   
Ford High School                      80.75         68.31%            79.30%   

                      Overall Passing Rate  
school_name                                 
Rodriguez High School               73.29%  
Figueroa High School                73.36%  
Huang High School                   73.50%  
Johnson High School                 73.64%  
Ford High School                    73.80%

## Math Scores by Grade

* Create a table that lists the average Reading Score for students of each grade level (9th, 10th, 11th, 12th) at each school.

  * Create a pandas series for each grade. Hint: use a conditional statement.
  
  * Group each series by school
  
  * Combine the series into a dataframe
  
  * Optional: give the displayed data cleaner formatting

In [47]:
# "Average Math Score" for each grade level (9th, 10th, 11th, 12th) at each school

math_table = school_data_complete[["school_name", "grade", "math_score"]]
# grade 9
math_9 = math_table.loc[math_table["grade"] == "9th"].groupby("school_name")
avg_math_9 = math_9["math_score"].mean()

# grade 10
math_10 = math_table.loc[math_table["grade"] == "10th"].groupby("school_name")
avg_math_10 = math_10["math_score"].mean()

# grade 11
math_11 = math_table.loc[math_table["grade"] == "11th"].groupby("school_name")
avg_math_11 = math_11["math_score"].mean()

#grade 12
math_12 = math_table.loc[math_table["grade"] == "12th"].groupby("school_name")
avg_math_12 = math_12["math_score"].mean()


math_scores = pd.DataFrame({
    "9th":avg_math_9,
    "10th":avg_math_10,
    "11th":avg_math_11,
    "12th":avg_math_12
})

math_scores["9th"] = math_scores["9th"].map("{:.2f}".format)
math_scores["10th"] = math_scores["10th"].map("{:.2f}".format)
math_scores["11th"] = math_scores["11th"].map("{:.2f}".format)
math_scores["12th"] = math_scores["12th"].map("{:.2f}".format)
math_scores

9th   10th   11th   12th
school_name                                      
Bailey High School     77.08  77.00  77.52  76.49
Cabrera High School    83.09  83.15  82.77  83.28
Figueroa High School   76.40  76.54  76.88  77.15
Ford High School       77.36  77.67  76.92  76.18
Griffin High School    82.04  84.23  83.84  83.36
Hernandez High School  77.44  77.34  77.14  77.19
Holden High School     83.79  83.43  85.00  82.86
Huang High School      77.03  75.91  76.45  77.23
Johnson High School    77.19  76.69  77.49  76.86
Pena High School       83.63  83.37  84.33  84.12
Rodriguez High School  76.86  76.61  76.40  77.69
Shelton High School    83.42  82.92  83.38  83.78
Thomas High School     83.59  83.09  83.50  83.50
Wilson High School     83.09  83.72  83.20  83.04
Wright High School     83.26  84.01  83.84  83.64

## Reading Score by Grade 

* Perform the same operations as above for reading scores

In [48]:
# "Average Reading Score" for each grade level (9th, 10th, 11th, 12th) at each school

read_table = school_data_complete[["school_name", "grade", "reading_score"]]
# grade 9
read_9 = read_table.loc[math_table["grade"] == "9th"].groupby("school_name")
avg_read_9 = read_9["reading_score"].mean()

# grade 10
read_10 = read_table.loc[math_table["grade"] == "10th"].groupby("school_name")
avg_read_10 = read_10["reading_score"].mean()

# grade 11
read_11 = read_table.loc[math_table["grade"] == "11th"].groupby("school_name")
avg_read_11 = read_11["reading_score"].mean()

#grade 12
read_12 = read_table.loc[math_table["grade"] == "12th"].groupby("school_name")
avg_read_12 = read_12["reading_score"].mean()


read_scores = pd.DataFrame({
    "9th":avg_read_9,
    "10th":avg_read_10,
    "11th":avg_read_11,
    "12th":avg_read_12
})
read_scores["9th"] = read_scores["9th"].map("{:.2f}".format)
read_scores["10th"] = read_scores["10th"].map("{:.2f}".format)
read_scores["11th"] = read_scores["11th"].map("{:.2f}".format)
read_scores["12th"] = read_scores["12th"].map("{:.2f}".format)
read_scores

9th   10th   11th   12th
school_name                                      
Bailey High School     81.30  80.91  80.95  80.91
Cabrera High School    83.68  84.25  83.79  84.29
Figueroa High School   81.20  81.41  80.64  81.38
Ford High School       80.63  81.26  80.40  80.66
Griffin High School    83.37  83.71  84.29  84.01
Hernandez High School  80.87  80.66  81.40  80.86
Holden High School     83.68  83.32  83.82  84.70
Huang High School      81.29  81.51  81.42  80.31
Johnson High School    81.26  80.77  80.62  81.23
Pena High School       83.81  83.61  84.34  84.59
Rodriguez High School  80.99  80.63  80.86  80.38
Shelton High School    84.12  83.44  84.37  82.78
Thomas High School     83.73  84.25  83.59  83.83
Wilson High School     83.94  84.02  83.76  84.32
Wright High School     83.83  83.81  84.16  84.07

## Scores by School Spending

* Create a table that breaks down school performances based on average Spending Ranges (Per Student). Use 4 reasonable bins to group school spending. Include in the table each of the following:
  * Average Math Score
  * Average Reading Score
  * % Passing Math
  * % Passing Reading
  * Overall Passing Rate (Average of the above two)

In [49]:
print(sch_summary_df['Per Student Budget'])

school_name
Bailey High School       $628.00
Cabrera High School      $582.00
Figueroa High School     $639.00
Ford High School         $644.00
Griffin High School      $625.00
Hernandez High School    $652.00
Holden High School       $581.00
Huang High School        $655.00
Johnson High School      $650.00
Pena High School         $609.00
Rodriguez High School    $637.00
Shelton High School      $600.00
Thomas High School       $638.00
Wilson High School       $578.00
Wright High School       $583.00
Name: Per Student Budget, dtype: object


In [50]:
# Sample bins. Feel free to create your own bins.
bins = [0, 585, 615, 645, 675]
group_names = ["<$585", "$585-615", "$615-645", "$645-675"]


scores_spending = sch_summary_0[["Average Math Score", 
                                  "Average Reading Score", 
                                  "% Passing Math", 
                                  "% Passing Reading", 
                                  "Overall Passing Rate"]].groupby(pd.cut(sch_summary_0["Per Student Budget"], bins=bins, labels=group_names )).mean()

scores_spending["Average Math Score"] = scores_spending["Average Math Score"].map("{:.2f}".format)
scores_spending["Average Reading Score"] = scores_spending["Average Reading Score"].map("{:.2f}".format)
scores_spending["% Passing Math"] = scores_spending[ "% Passing Math"].map("{:.2f}%".format)
scores_spending["% Passing Reading"] = scores_spending[ "% Passing Reading"].map("{:.2f}%".format)
scores_spending["Overall Passing Rate"] = scores_spending["Overall Passing Rate"].map("{:.2f}".format)
scores_spending.head()

Average Math Score Average Reading Score % Passing Math  \
Per Student Budget                                                           
<$585                           83.46                 83.93         93.46%   
$585-615                        83.60                 83.89         94.23%   
$615-645                        79.08                 81.89         75.67%   
$645-675                        77.00                 81.03         66.16%   

                   % Passing Reading Overall Passing Rate  
Per Student Budget                                         
<$585                         96.61%                95.04  
$585-615                      95.90%                95.07  
$615-645                      86.11%                80.89  
$645-675                      81.13%                73.65

## Scores by School Size

* Perform the same operations as above, based on school size.

In [51]:
# Sample bins. Feel free to create your own bins.
size_bins = [0, 1000, 2000, 5000]
group_names = ["Small (<1000)", "Medium (1000-2000)", "Large (2000-5000)"]


scores_size = sch_summary_0[["Average Math Score", 
                                  "Average Reading Score", 
                                  "% Passing Math", 
                                  "% Passing Reading", 
                                  "Overall Passing Rate"]].groupby(pd.cut(sch_summary_0["Total Students"], bins=size_bins, labels=group_names )).mean()


scores_size["Average Math Score"] = scores_size["Average Math Score"].map("{:.2f}".format)
scores_size["Average Reading Score"] = scores_size["Average Reading Score"].map("{:.2f}".format)
scores_size["% Passing Math"] = scores_size[ "% Passing Math"].map("{:.2f}%".format)
scores_size["% Passing Reading"] = scores_size[ "% Passing Reading"].map("{:.2f}%".format)
scores_size["Overall Passing Rate"] = scores_size["Overall Passing Rate"].map("{:.2f}".format)
scores_size.head()


Average Math Score Average Reading Score % Passing Math  \
Total Students                                                               
Small (<1000)                   83.82                 83.93         93.55%   
Medium (1000-2000)              83.37                 83.86         93.60%   
Large (2000-5000)               77.75                 81.34         69.96%   

                   % Passing Reading Overall Passing Rate  
Total Students                                             
Small (<1000)                 96.10%                94.82  
Medium (1000-2000)            96.79%                95.20  
Large (2000-5000)             82.77%                76.36

## Scores by School Type

* Perform the same operations as above, based on school type.

In [53]:
scores_type = sch_summary_0[["School Type", "Average Math Score", 
                                  "Average Reading Score", 
                                  "% Passing Math", 
                                  "% Passing Reading", 
                                  "Overall Passing Rate"]]
scores_type = scores_type.groupby("School Type").mean()


scores_type["Average Math Score"] = scores_type["Average Math Score"].map("{:.2f}".format)
scores_type["Average Reading Score"] = scores_type["Average Reading Score"].map("{:.2f}".format)
scores_type["% Passing Math"] = scores_type[ "% Passing Math"].map("{:.2f}%".format)
scores_type["% Passing Reading"] = scores_type[ "% Passing Reading"].map("{:.2f}%".format)
scores_type["Overall Passing Rate"] = scores_type["Overall Passing Rate"].map("{:.2f}".format)
scores_type.head()

Average Math Score Average Reading Score % Passing Math  \
School Type                                                           
Charter                  83.47                 83.90         93.62%   
District                 76.96                 80.97         66.55%   

            % Passing Reading Overall Passing Rate  
School Type                                         
Charter                96.59%                95.10  
District               80.80%                73.67

### Analysis of PyCitySchools

Schools who have $615 or more dollars per student are actually underperforming their peers who are at school with lower per student budgets. 

Schools with a large population, (2000-5000) students, show there is a difference in performance compared to the low and medium enrolled schools. Schools with a large population are underperforming overall, with math and reading percentages also clearly below the other types of schools.  

The type of school shows that Charter schools have a 94% passing math percentage compared to the Distict schools which only has a 67% passing rate. 